### ISSUE TITLE
cache expire issue

### ISSUE THREAD

**Issue by leoliu** [fogfish/cache#23]
Run the following eshell:

```
1> cache:start_link(c, [{n,10}, {ttl, 60}]).
2> cache:put(c, x, test, 10).
```

What is the upper bound ttl of x in cache? I thought it should not be more than 16 seconds but somehow x stays there for minutes. Is this a bug?

**Comment by fogfish** [comment]
Thank you for the report!

I would say this is an issue with documentation. It seems that I have to make a dedicated write-up about some of the feature.

Please keep in-mind the following cache feature:

* The chosen segment-based solution has a cons - inability to assign precise TTL per single cache entry. TTL is always approximated to nearest segment. (e.g. cache with 60 sec TTL and 10 segments has 6 sec accuracy on TTL).

* `cache:get(...)` always prolongs value ttl and it prolongs it to max TTL value. 

```
cache:put(c, x, test, 10).   // x TTL in range of 10-16 sec
timer:sleep(9000).
cache:get(c, x).                  // x TTL in range of 60 sec
timer:sleep(50000).
cache:get(c, x).                  // x TTL in range of 60 sec
```

* There is a feature `cache:lookup(...)` the operation do not prolongs ttl

```
cache:put(c, x, test, 10).   // x TTL in range of 10-16 sec
timer:sleep(9000).
cache:lookup(c, x).           // you'll get test
timer:sleep(9000).
cache:lookup(c, x).           // you'll get undefined
```

This behaviour (might) looks *wired* if you compare with memcache/redis. They do have a dedicated `touch` routine to prolong ttl (`cache:get(...)` is equivalent of `memcache:touch(...)`).  I'll be glad to continue discussion with you about auto ttl-prolong feature.

However, you report helped me to spot one clear bug about `cache:ttl(...)`. It returns wrong value.

**Comment by leoliu** [comment]
Thanks a lot for the detailed explanation. I just checked my code from a year ago and it was using `cache:lookup/2`. But the eshell session I tried for this report was using `cache:get/2`. Somehow I forgot this subtle difference that the latter prolongs ttl. But I opened this issue because of noticing something in cache significantly longer than intended. So I tried this example:

```
cache:put(c, x, test, 10).
timer:sleep(9000).
cache:lookup(c, x).
timer:sleep(9000).
cache:lookup(c, x).
```

I got `test` from both lookups. It only returned `undefined` after ~30s.

**Comment by fogfish** [comment]
This is unfortunately bug (or feature). Existed implementation starts to evict segments when heap if full.
https://github.com/fogfish/cache/blob/master/src/cache_bucket.erl#L221

I've been trying to find a compromise between lazy initialisation vs ttl management. It would be much easy if all heap segments are created when `cache:start_link(...)` is called because all segments in place and you can put value to right one. The existed implementation used lazy (premature optimisation) of creating a new segments on the demand. Thus, initially cache badly obeys ttl.

I think I need to fix this one.

**Comment by leoliu** [comment]
BTW, while you are at it the type for `#cache.policy` is not `integer()`.

**Comment by fogfish** [comment]
FYI: I've opened PR #25 but I've not run yet deep integration testing. I'll try to improve this PR during weekend. I'll appreciate if you have ability to test it with your code. I am targeting to merge it and release hex package by next Mon/Tue.

**Comment by leoliu** [comment]
I am running the `explicit-segment-allocation` branch and the issue seems fixed. Thanks for the work. I'll continue using this branch and post back if something happens. There are two compiler warnings:

```
_build/default/lib/cache/src/cache_heap.erl:104: Warning: variable 'B' is unused
_build/default/lib/cache/src/cache_bucket.erl:467: Warning: variable 'Val' is unused
```

**Comment by leoliu** [comment]
I have one question. Given the precision of 6 seconds (`[{n,10}, {ttl, 60}]`), is it guaranteed x will live for at least 10 seconds in the following example? Thanks.

```
cache:put(c, x, test, 10).
```

**Comment by fogfish** [comment]
closed by #25
